design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,480 @@
+*  This design doesn't include the creation of the destination PVCs, it expects the destination PVCs to exist.
+
+ ## Definition of Users
+This feature is for all KubeVirt users who want to move the PVCs assigned to a VM without workload interruptions.
+
+## User Stories
+
+  1. As a VM owner, the storage class of volumes used by a running VM is deprecated, and the volumes need to be moved to another storage class.
+  2. As a VM owner, I want to use higher performance storage class that is now available in the cluster.
+  3. As a VM owner, I see my non-resizable storage device filling up and I want to move my VM to a larger storage device without interruptions in order to give it more storage.
+  4. As an infra admin, I need to replace the storage array due to a hardware refresh and I want to do this without stopping the VMs in order to make this infra operation transparent to VM owners.
+
+## Repos
+
+  - [kubevirt/kubevirt](https://github.com/kubevirt/kubevirt)
+
+# Design
+This proposal introduces a new CRD called `VolumeMigration`
+
+## Volume migration custom resource
+
+```golang
+type VolumeMigration struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+	Spec VolumeMigrationSpec `json:"spec"valid:"required"`
+	Status VolumeMigrationStatus `json:"status,omitempty"`
+}
+
+type VolumeMigrationList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items []VolumeMigration `json:"items"`
+}
+
+type SourcePvcReclaimPolicy string
+const (
+	SourcePvcReclaimPolicyDelete SourcePvcReclaimPolicy = "Delete"
+	SourcePvcReclaimPolicyRetain SourcePvcReclaimPolicy = "Retain"
+)
+
+type MigratedVolume struct {
+	SourcePvc      string `json:"sourcePvc" valid:"required"`
+	DestinationPvc string `json:"destinationPvc" valid:"required"`
+	SourcePvcReclaimPolicy SourcePvcReclaimPolicy `json:"sourcePvcReclaimPolicy,omitempty"`
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRunning    = "Running"
+	VolumeMigrationPhaseScheduling = "Scheduling"
+	VolumeMigrationPhaseSucceeded  = "Succeeded"
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhaseUnknown    = "Unknown"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type MigratedVolumeState string
+
+const (
+        MigratedVolumeStateValid    = "Valid"
+        MigratedVolumeStatePending  = "Pending"
+        MigratedVolumeStateRejected = "Rejected"
+)
+
+type VolumeMigrationState struct {
+        MigratedVolume `json:",inline"`
+        State     MigratedVolumeState `json:"state,omitempty"`
+        Reason    *string             `json:"reason,omitempty"`
```

Comment:
Wondering if a client of this API may like more granular rejected states?  For example, some volumes will get rejected for intrinsic reasons like it's virtiofs, others for temporary reasons like hotplug, others for implementation details that may change later like multiple VMIs

Reply:
@mhenriks I don't quite understand your comment. The reason applies for a single migrated volume pair. If you look at the third example in the [API example](https://github.com/alicefr/community/blob/storage-class-migration/design-proposals/volume-migration/volume-migration.md#api-examples), there are 2 rejected volumes for 2 distinct reasons.

Reply:
Ok, in the actual code, I have declared those [strings](https://github.com/alicefr/kubevirt/blob/storage-migration/staging/src/kubevirt.io/api/storage/v1alpha1/types.go#L71-L78) but I can create a custom type for the reason if you prefer. For the VolumeStatus, we are also using [a string for `Reason`](https://github.com/kubevirt/kubevirt/blob/main/staging/src/kubevirt.io/api/core/v1/types.go#L335)